### PR TITLE
fix #856: block envs set in resources/AppRun when invoking appstream utilities

### LINKED ISSUE #856: AppStream validation fails with symbol lookup error
When I run appimagetool on my AppDir without `--no-appstream`, I get the following error:

```
AppStream upstream metadata found in usr/share/metainfo/org.keepassxc.KeePassXC.appdata.xml
Trying to validate AppStream information with the appstreamcli tool
In case of issues, please refer to https://github.com/ximion/appstream
/bin/appstreamcli: symbol lookup error: /usr/lib/libappstream.so.4: undefined symbol: g_strv_contains
Failed to validate AppStream information with appstreamcli
```

This seems to be a glib incompatibility. appstreamcli is not bundled with the AppImage, so both the CLI tool as well as the the failing lib are from my system, but it works when I use it without the appimagetool libs in my library load path.

I'm on Arch Linux.

### PR DESCRIPTION
**Pull request by omegacoleman**
This fixes the bug described in:

https://github.com/AppImage/AppImageKit/issues/856
https://github.com/AppImage/AppImageKit/issues/1048
https://github.com/linuxdeploy/linuxdeploy-plugin-appimage/issues/14
https://github.com/linuxdeploy/linuxdeploy-plugin-appimage/issues/16
https://github.com/AppImage/AppImageKit/issues/562

the original bug was caused by appimagetool using `system(3)` to execute a utility installed on the host system (say, /usr/bin/appstreamcli). `system(3)` passes `LD_LIBRARY_PATH` which causing the utiliy to load the wrong library, and break.

This PR solve the problem by dissolving `system(3)` into `fork(2)` `execve(2)` and `waitpid(2)`, and use `unsetenv` on `LD_LIBRARY_PATH` and other envs set in resources/AppRun in the child process to fix the bug.

as discussed in #856 , the ultimate solution could be banishing the modification of `LD_LIBRARY_PATH`, however, as this issue is already 2 years old, and everybody's punning the `LD_LIBRARY_PATH`, and AppRun.c provided here is still modifying it, I propose accept this PR as a fix.

### CODE REVIEW

**Comment by omegacoleman** [comment]
@probonopd @TheAssassin

**Comment by TheAssassin** [comment]
A simple `unsetenv()` would've been easier. Just saying.

The main issue with your PR is that setting `$LD_LIBRARY_PATH` might be set by the caller intentionally. We don't bundle the tool. Hence, we shouldn't make the assumption that the one in `$PATH` works as-is without a user-specified variable.

The true solution is to actually bundle appimagetool properly, e.g., with linuxdeploy. I proposed this long ago, but it was rejected immediately. The bundling performed in the shell script mess (that I had to touch again recently) is just that: a mess.

Getting rid of that old, known-broken `AppRun.c` could be considered a solution. This is just another workaround with major side effects. The discussion therefore centers around "are these side effects better than the current ones". I tend to agree. And given that this tool will likely be replaced in the future by one written in a better programming language than C (which is easier to get right), I would say, we could merge this PR. But note that this is more of a "better than before" than a "yay, it's finally solved".

I haven't had a look at the code yet. I think it's adding complexity unnecessarily, and will take more than a minute to review. Perhaps using `unsetenv()` somewhere in `main()` is easier. What do you think?

**Comment by probonopd** [comment]
>The true solution is to actually bundle appimagetool properly

Nothing against this as long as it can be done within the AppImageKit repo without introducing dependencies on external tools, and even more so ones not in the AppImage organization.

**Comment by omegacoleman** [comment]
> Edit: we get notifications, buddy. Don't worry. Pinging is annoying, it generates yet another mail, and is kinda discouraging. Especially if it's done after like just 10 minutes.

fine, actually there's plenty of project on github which requires PM to assign an assignee of related module before doing anything with the PR, and what I'm doing is to suggest candidates so that this process got faster, not meaning 'lookit this fast'

p.s. no one would know your habits ahead if you don't put that in CONTRIBUTING.md

> A simple `unsetenv()` would've been easier. Just saying.
> 
> The main issue with your PR is that setting `$LD_LIBRARY_PATH` might be set by the caller intentionally. We don't bundle the tool. Hence, we shouldn't make the assumption that the one in `$PATH` works as-is without a user-specified variable.
> I haven't had a look at the code yet. I think it's adding complexity unnecessarily, and will take more than a minute to review. Perhaps using `unsetenv()` somewhere in `main()` is easier. What do you think?
> 

I've not tested that but there's a few other `exec*()` which calls bundled tools, like `zsyncmake`. If those programs rely on `$LD_LIBRARY_PATH`, then unsetting it cause the bundled programs to erroneously link to host libraries.

> 
> The true solution is to actually bundle appimagetool properly, e.g., with linuxdeploy. I proposed this long ago, but it was rejected immediately. The bundling performed in the shell script mess (that I had to touch again recently) is just that: a mess.
> 
> Getting rid of that old, known-broken `AppRun.c` could be considered a solution. This is just another workaround with major side effects. The discussion therefore centers around "are these side effects better than the current ones". I tend to agree. And given that this tool will likely be replaced in the future by one written in a better programming language than C (which is easier to get right), I would say, we could merge this PR. But note that this is more of a "better than before" than a "yay, it's finally solved".
> 

I agree with the design but I think you already missed the chance. There are already tons of apps with their AppRun's modifying the `LD_LIBRARY_PATH`, and most of them don't invoke host executables, so it's hard to persuade them all to shift to the new version. Even if you managed to fix the AppRun script, next time an app using old measures bundling AppImageKit will crash again.

And for linuxdeploy, it could introduce a circular-dependency issue as it bundles AppImageKit to run. Anyway, I would really appreciate that if you successfully solve the mess. Or just learn the way of UNIX -- we live on with our sins lol.

**Comment by TheAssassin** [comment]
> I've not tested that but there's a few other exec*() which calls bundled tools, like zsyncmake. If those programs rely on $LD_LIBRARY_PATH, then unsetting it cause the bundled programs to erroneously link to host libraries.

Right, I confused something there. I actually thought about adding that in the subprocess you create with `fork()`. When you're in the child, you can do whatever you want in the environment, it should not have an influence on the parent process.

**Comment by omegacoleman** [comment]
> > I've not tested that but there's a few other exec*() which calls bundled tools, like zsyncmake. If those programs rely on $LD_LIBRARY_PATH, then unsetting it cause the bundled programs to erroneously link to host libraries.
> 
> Right, I confused something there. I actually thought about adding that in the subprocess you create with `fork()`. When you're in the child, you can do whatever you want in the environment, it should not have an influence on the parent process.

This is applied. I also checked other envs set in `resources/AppRun` which should be the devious shell script you mentioned, and figured that it would even cause `appstream-util`'s glib to load the wrong gsettings schema, so I changed to unset them all to fix that problem too.

File: src/appimagetool.c [hunk 1]
```diff
@@ -794,19 +831,29 @@ main (int argc, char *argv[])
...
                     g_print("In case of issues, please refer to https://github.com/ximion/appstream\n");
-                    int ret = system(command);
+                    int ret = run_external(g_find_program_in_path ("appstreamcli"), args);
                     if (ret != 0)
                         die("Failed to validate AppStream information with appstreamcli");
                 }
                 /* It seems that hughsie's appstream-util does additional validations */
                 if(g_find_program_in_path ("appstream-util")) {
-                    sprintf (command, "%s validate-relax %s", g_find_program_in_path ("appstream-util"), appdata_path);
+                    char *args[] = {
```
Comment (TheAssassin) [inline src/appimagetool.c:848]: This is a great enhancement, as the previous version was obviously unsafe.

File: src/appimagetool.c [hunk 2]
```diff
@@ -457,6 +459,41 @@ bool readFile(char* filename, int* size, char** buffer) {
     return TRUE;
 }
 
+/* run a command outside the current appimage, block environs like LD_LIBRARY_PATH */
+int run_external(const char *filename, char *const argv []) {
+    int pid = fork();
+    if (pid == -1) {
```
Comment (TheAssassin) [inline src/appimagetool.c:465]: I think you could/should check `<= 0` here.

File: src/appimagetool.c [hunk 3]
```diff
@@ -457,6 +459,41 @@ bool readFile(char* filename, int* size, char** buffer) {
...
+    } else if (pid == 0) {
+        // blocks env defined in resources/AppRun
+        unsetenv("LD_LIBRARY_PATH");
+        unsetenv("PYTHONPATH");
+        unsetenv("XDG_DATA_DIRS");
+        unsetenv("PERLLIB");
+        unsetenv("GSETTINGS_SCHEMA_DIR");
+        unsetenv("QT_PLUGIN_PATH");
+        // runs command
+        execve(filename, argv, environ);
```
Comment (TheAssassin) [inline src/appimagetool.c:478]: You don't need to pass `environ` yourself, unless you want to specify your own, custom one. The parent environment will be used for child processes automatically.

File: src/appimagetool.c [hunk 4]
```diff
@@ -457,6 +459,41 @@ bool readFile(char* filename, int* size, char** buffer) {
     return TRUE;
 }
 
+/* run a command outside the current appimage, block environs like LD_LIBRARY_PATH */
+int run_external(const char *filename, char *const argv []) {
+    int pid = fork();
+    if (pid == -1) {
+        g_print("run_external: fork failed");
+        // fork failed
+        return -1;
```
Comment (TheAssassin) [inline src/appimagetool.c:468]: This is a fatal error, one should _always_ be able to `fork()`. You can `exit` here.

File: src/appimagetool.c [hunk 5]
```diff
@@ -457,6 +459,41 @@ bool readFile(char* filename, int* size, char** buffer) {
...
+    } else {
+        int wstatus;
+        if (waitpid(pid, &wstatus, 0) == -1) {
+            g_print("run_external: wait failed");
+            return -1;
+        }
+        if (WIFEXITED(wstatus) && (WEXITSTATUS(wstatus) == 0)) {
+            return 0;
+        } else {
+            g_print("run_external: subprocess exited with non-0 status");
```
Comment (TheAssassin) [inline src/appimagetool.c:491]: Please print and return the exact exit code here. This is not necessarily helpful to the user who must debug the subprocess failure.

File: src/appimagetool.c [hunk 6]
```diff
@@ -61,6 +61,8 @@ extern char runtime[];
 extern unsigned int runtime_len;
 #endif
 
+extern char **environ;
```
Comment (TheAssassin) [inline src/appimagetool.c:64]: You shouldn't define this on your own; instead, you could import it from an included header. However, you don't need it at all (see below).

**Review by TheAssassin — CHANGES_REQUESTED** [review]

**Comment by TheAssassin** [comment]
Looks much better and much less complex already, thanks. I've left some feedback.

File: src/appimagetool.c (hunk 2 above)
Comment (omegacoleman) [reply src/appimagetool.c]: `< 0` I suppose?

**Comment by omegacoleman** [comment]
The requested changes were made and rebased master.

**Review by TheAssassin — CHANGES_REQUESTED** [review]
Well, there's a few minor things left I'd change (code style), but it's good enough now. Thanks.

**Review by TheAssassin — APPROVED** [review]

**Comment by probonopd** [comment]
Could this be causing https://github.com/AppImage/AppImageKit/issues/1126?

**Comment by omegacoleman** [comment]
> Could this be causing #1126?

No.

If there's bug in the user's host installation, like python, glib, etc, this fix could trigger that though, but that's another story.

**Pull request merged**